core/block1.c

```diff
@@ -0,0 +1,141 @@
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+
+coap_status_t coap_block1_handler(lwm2m_context_t * contextP,
+                                  uint16_t mid,
+                                  uint8_t * buffer,
+                                  size_t length,
+                                  lwm2m_server_t * server,
+                                  uint16_t blockSize,
+                                  uint32_t blockNum,
+                                  bool blockMore,
+                                  uint8_t ** outputBuffer,
+                                  size_t * outputLength)
+{
+    if (server == NULL)
+    {
+        return COAP_500_INTERNAL_SERVER_ERROR;
+    }
+
+    // find current block1 data for this server
+    lwm2m_block1_data_t * block1Data = (lwm2m_block1_data_t *) LWM2M_LIST_FIND(contextP->block1DataList, server->shortID);
+
+
+    // manage new block1 transfer
+    if (blockNum == 0)
+    {
+       // we already have block1 data for this server, clear it
+       if (block1Data != NULL)
+       {
+           lwm2m_free(block1Data->block1buffer);
+       }
+       else
+       {
+           block1Data = lwm2m_malloc(sizeof(lwm2m_block1_data_t));
+           if (NULL == block1Data) return COAP_500_INTERNAL_SERVER_ERROR;
+           block1Data->serverID = server->shortID;
+           contextP->block1DataList = (lwm2m_block1_data_t *) LWM2M_LIST_ADD(contextP->block1DataList, block1Data);
+       }
+
+       block1Data->block1buffer = lwm2m_malloc(length);
+       block1Data->block1bufferSize = length;
+
+       // write new block in buffer
+       memcpy(block1Data->block1buffer, buffer, length);
+       block1Data->lastmid = mid;
+    }
+    // manage already started block1 transfer
+    else
+    {
+       if (block1Data == NULL)
+       {
+           // we never receive the first block
+           // TODO should we clean block1 data for this server ?
+           return COAP_408_REQ_ENTITY_INCOMPLETE;
+       }
+
+       // If this is a retransmission, we already do that.
+       if (block1Data->lastmid != mid)
+       {
+           if (block1Data->block1bufferSize != blockSize * blockNum)
+          {
+              // we don't receive block in right order
+              // TODO should we clean block1 data for this server ?
+              return COAP_408_REQ_ENTITY_INCOMPLETE;
+          }
+
+          // re-alloc new buffer
+          uint8_t * oldBuffer = block1Data->block1buffer;
+          size_t oldSize = block1Data->block1bufferSize;
+          block1Data->block1bufferSize = oldSize+length;
+          block1Data->block1buffer = lwm2m_malloc(block1Data->block1bufferSize);
+          if (NULL == block1Data->block1buffer) return COAP_500_INTERNAL_SERVER_ERROR;
+          memcpy(block1Data->block1buffer, oldBuffer, oldSize);
+          lwm2m_free(oldBuffer);
+
+          // write new block in buffer
+          memcpy(block1Data->block1buffer + (blockNum * blockSize), buffer, length);
```

Comment:
maybe using "oldSize" instead of "(blockNum * blockSize)" is nicer, but not really required.